pkg/utils/grpc.go

```diff
@@ -27,44 +46,139 @@ func NewGrpc(serverType string, port int, certPem, keyPem, rootCertPem []byte) (
 
 	log.Debug().Msgf("Parameters for %s gRPC server: MaxConcurrentStreams=%d;  KeepAlive=%+v", serverType, maxStreams, streamKeepAliveDuration)
 
+	s := &Server{
+		name:     serverType,
+		cm:       cm,
+		certName: certName,
+	}
+
 	grpcOptions := []grpc.ServerOption{
 		grpc.MaxConcurrentStreams(maxStreams),
 		grpc.KeepaliveParams(keepalive.ServerParameters{
 			Time: streamKeepAliveDuration,
 		}),
 	}
 
-	mutualTLS, err := setupMutualTLS(false, serverType, certPem, keyPem, rootCertPem)
-	if err != nil {
-		log.Error().Err(err).Msg("Error setting up mutual tls for GRPC server")
-		return nil, nil, err
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: false,
+		ServerName:         serverType,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+		MinVersion:         constants.MinTLSVersion,
+		GetConfigForClient: func(*tls.ClientHelloInfo) (*tls.Config, error) {
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			return &s.config, nil
+		},
 	}
+	mutualTLS := grpc.Creds(credentials.NewTLS(&tlsConfig))
 	grpcOptions = append(grpcOptions, mutualTLS)
 
-	return grpc.NewServer(grpcOptions...), lis, nil
+	s.server = grpc.NewServer(grpcOptions...)
+	return s, lis, nil
+}
+
+// GetServer returns the gRPC server
+func (s *Server) GetServer() *grpc.Server {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.server
+}
+
+func (s *Server) setConfig() error {
+	grpcCert, err := s.cm.IssueCertificate(
+		s.certName,
+		certificate.Internal,
+		certificate.FullCNProvided())
+	if err != nil {
+		return err
+	}
+
+	certif, err := tls.X509KeyPair(grpcCert.GetCertificateChain(), grpcCert.GetPrivateKey())
+	if err != nil {
+		return fmt.Errorf("[grpc][mTLS][%s] Failed loading Certificate (%+v) and Key (%+v) PEM files", s.name, grpcCert.GetCertificateChain(), grpcCert.GetPrivateKey())
+	}
+
+	certPool := x509.NewCertPool()
+
+	// Load the set of Root CAs
+	if ok := certPool.AppendCertsFromPEM(grpcCert.GetTrustedCAs()); !ok {
+		return fmt.Errorf("[grpc][mTLS][%s] Failed to append client certs", s.name)
+	}
+
+	s.mu.Lock()
+	s.config = tls.Config{
+		InsecureSkipVerify: false,
+		ServerName:         s.name,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+		Certificates:       []tls.Certificate{certif},
+		ClientCAs:          certPool,
+		MinVersion:         constants.MinTLSVersion,
+	}
+	s.mu.Unlock()
+
+	return nil
+}
+
+func (s *Server) configureConfigUpdates(ctx context.Context) error {
+	// listen for certificate rotation first, so we don't miss any events
+	certRotationChan, unsubscribeRotation := s.cm.SubscribeRotations(s.certName)
+
```

Comment:
Add `defer unsubscribeRotation()` and drop other unsubscriptions below.

Reply:
Does `unsubscribeRotation()` have to be called in the go routine? If it is called outside of the go routine then we'll unsubscribe from the pubsub while we're still listening for rotation events.

Reply:
In that case, would it make sense to only subscribe within the goroutine? Alternatively,  It might be more readable to let the caller invoke this as `go watchForConfigUpdates()` and drop the wrapping of the goroutine in this function, but it seems like you may be relying on some sort of error propagation down the call stack to terminate the process when this fails to happen. I think a log.Fatal() should suffice in such cases, or by passing a signal handler from main to such functions, vs structuring the code such that a lot of function calls in the call stack need to propagate the error all the way back to `main()`.

Reply:
We try to subscribe to certificate rotations prior to issuing the certificate to ensure that rotations are not missed between issuing and subscribing. However, this is an edge case

Reply:
Got it. Would be nice to update the comment in L124 to indicate that the issuance happens via the call to `initConfig()`